// wechat_priv/lib/wechat_priv.js
var socket_arr = require('../../socket_arr');
var m_user_query = require('./user_query');
var m_user_set = require('./user_set');
/*var pg = require('pg');
var db = require('db');
var pgString = db.consqlString;
var client = new pg.Client(pgString);
client.connect();

*/


var mysql = require('mysql');
var client = mysql.createConnection({
    host   : 'localhost',
    user   : 'root',
    password : '123456',
    database : 'cloud_server_db',
    port : '3306'
});
 
client.connect();

var weixin_socket_arr    = new Array();

var user_row_string = "";

function db_res_bind_select_query(select_result)
{
    console.log("Enter db_res_bind_select_query");

    if ( 0 == select_result.rowCount)
    {
        var insert_string = "INSERT INTO user_table VALUES ('"
                    + this.userID + "','"
                    + this.deviceID + "','"
                    + 1 + "')";

        console.log(insert_string);

        var insert_query = client.query(insert_string);
    }
    else
    {
        console.log(" 0 != row");
    }
}

function message_handle(req, res)
{      
    
    var message = req.weixin;

    console.log(message);  

    var msgType = message.MsgType;

    var eventKey = message.EventKey;

    var weixin_user = message.FromUserName;

    console.log(msgType);

    if ( "device_event" == msgType)
    {       
        console.log("Enter device_event");
     
        console.log("event = " + message.Event);

        var userID = message.OpenID;

	var deviceID = message.DeviceID;

        if ( "bind" == message.Event)
        {
            console.log("Enter Event bind");
            var select_string = "SELECT * FROM user_table WHERE user_id = '" 
                                + userID + "' and device_id = '" + deviceID + "'";

            console.log(select_string);

            var select_query = client.query( select_string);
            select_query.userID   = userID;
            select_query.deviceID = deviceID;
            select_query.on('end', db_res_bind_select_query);
        }

        else if ( "unbind" == message.Event)
        {
            console.log("Enter Event unbind");
 
            var delete_string = "DELETE FROM user_table WHERE user_id = '"
                              + userID + "' and device_id = '" + deviceID + "'";
            var delete_query  = client.query(delete_string);
        }
    }

    else if ( "event" == msgType)
    {
	if (("query" == eventKey)
	    || ("turn_on" == eventKey)
	    || ("turn_off" == eventKey)
	   )
	{

            console.log("Enter weixin");
            req.socket.id = weixin_user;
            req.socket.res = res;
            socket_arr.socket_arr_add(weixin_socket_arr, req.socket, weixin_user);

            if ("query" == eventKey)
            {

                var query_string = "SELECT * FROM user_table WHERE user_id = '" + weixin_user + "';" ;
                var user_query = client.query(query_string);
                user_query.user_id = weixin_user;
                user_query.res = res;

                user_query.user_rows = new Array();
                user_query.on('row', m_user_query.user_row_result);
                user_query.on('end', m_user_query.user_end_result);
                   
              /*
                res.reply({
                  type: 'text',
                  content: 'query'
                }); */

            }

            else if ( "turn_on" == eventKey)
            {

                var query_string = "SELECT * FROM user_table WHERE user_id = '" + weixin_user + "';" ;
                var user_query = client.query(query_string);
                user_query.user_id = weixin_user;
                user_query.res = res;
                user_query.cmd = "turn_on";

                user_query.user_rows = new Array();
                user_query.on('row', m_user_set.user_row_result);
                user_query.on('end', m_user_set.user_end_result);

/*
                res.reply({
                  type: 'text',
                  content: 'turn on successfully'
                });
*/
            }

            else 
            {

                var query_string = "SELECT * FROM user_table WHERE user_id = '" + weixin_user + "';" ;
                var user_query = client.query(query_string);
                user_query.user_id = weixin_user;
                user_query.res = res;
                user_query.cmd = "turn_off";

                user_query.user_rows = new Array();
                user_query.on('row', m_user_set.user_row_result);
                user_query.on('end', m_user_set.user_end_result);
            }

            req.socket.on('close', weixin_close_socket);
        }
    }

    else
    {
    }

}


function weixin_close_socket()
{
    socket_arr.socket_arr_del(weixin_socket_arr, this.id);
}

module.exports.message_handle = message_handle;
module.exports.weixin_socket_arr = weixin_socket_arr;
