// dev_conn/dev_conn.js
var mysql   = require('mysql');
var db = require('db');
var client = mysql.createConnection(db.mysql);
 
client.connect();





var g_connId=1;
var sockets = {};
var devConnMap = {};
var devIdArry=new Array();

function setDevOffline(arr) {
    var devIdListStr = "";

    console.log("setDevOffline");

  /* if(arr.length == 0) {
        return;
    }*/

    for(var i=0; i<arr.length; i++) {
        devIdListStr = devIdListStr + arr[i];
        if(i<arr.length-1) {
            devIdListStr = devIdListStr + ",";
        }
    }
    console.log("devIdListStr:"+devIdListStr);

    var updateStr="UPDATE iot_device SET online='0' WHERE deviceId IN (" + devIdListStr + ")";

    console.log("setDevOffline updateStr="+updateStr);
    client.query(updateStr, function(err, result) { 
        if (err) {
            console.error(err.stack);
        }
    });
}

function newConnection(socket) {
    var skInfo = {};
    skInfo.sk=socket;
    skInfo.devList = new Array();
    sockets[g_connId] = skInfo;
    return g_connId++;
}

function releaseConnection(connId) {
    if(typeof(connId) == 'undefined') {
        console.log("releaseConnection connId is undefined!");
        return;
    }

    console.log("releaseConnection connId:"+connId);
    //console.log(sockets[connId]);
 /*   if(typeof(sockets[connId]) == 'undefinded') {
        console.log("releaseConnection missing info for connId " + connId + "in sockets!");
        return;
    }
*/
    //setDevOffline(sockets[connId].devList);
    //delete sockets[connId];
    setDevOffline(devIdArry);
}

function releaseConnections(connId) {
    if(typeof(connId) == 'undefined') {
        console.log("releaseConnection connId is undefined!");
        return;
    }

    console.log("releaseConnection connId:"+connId);
    //console.log(sockets[connId]);
 /*   if(typeof(sockets[connId]) == 'undefinded') {
        console.log("releaseConnection missing info for connId " + connId + "in sockets!");
        return;
    }
*/
    //setDevOffline(sockets[connId].devList);
    delete sockets[connId];
    //setDevOffline(devIdArry);
}




function updateDevInfo(devId, connId) {
    //devConnMap[devId] = connId;
    console.log("devId is :"+devId); 
    console.log(connId);

  /*  if(typeof(sockets[connId]) == 'undefinded') {
        console.log("updateDevInfo missing info for connId " + connId + "in sockets!");
        return;
    }*/
    //sockets[connId].devList.push(devId);
    devIdArry.push(devId);
    var devIdListStr = "";
    for(var i=0; i<devIdArry.length; i++) {
        devIdListStr = devIdListStr + devIdArry[i];
        if(i<devIdArry.length-1) {
            devIdListStr = devIdListStr + ",";
        }
    }
    console.log("devIdListStr:"+devIdListStr);
}

function getDevListByConnId(connId) {
    return sockets[connId].devList;
}

function getSocketByDevId(devId) {
    var connId = devConnMap[devId];
    if(typeof(connId) == 'undefined') {
        console.log("Can't find connId for devId" + devId);
        return -1;
    }

    if(typeof(sockets[connId]) == 'undefined') {
        console.log("getSocketByDevId missing info for connId " + connId + "in sockets!");
        return;
    }
    
    if(typeof(sockets[connId].sk) == 'undefined') {
        console.log("getSocketByDevId missing info for connId " + connId + ".sk in sockets!");
        return;
    }

    return sockets[connId].sk;
}

exports.newConnection = newConnection;
exports.releaseConnection = releaseConnection;
exports.releaseConnections = releaseConnections;
exports.updateDevInfo = updateDevInfo;
exports.getDevListByConnId = getDevListByConnId;
exports.getSocketByDevId = getSocketByDevId;
